### ISSUE TITLE
cant undeploy a construction conyard  [$20]

### ISSUE THREAD

**Issue by MicroJOo** [OpenRA/OpenRA#8041]
okay , here is the deal ... 

i have experienced in the new release (1 time so far) , that i couldnt undeploy a construction conyard (my 2nd one to be precise) ... all i know is , it was under attack , and ive tried to undeploy during the attack ... 

later on , ive investigated the replay , but couldnt find any indicator ... started a debug-game and tried various things to reproduce it again ... but it didnt show up again .... 

within 2 days , several people reported this problem in the lobby ... 

![undeploy](https://cloud.githubusercontent.com/assets/9139965/7339121/4f86e2a2-ec63-11e4-87c6-a071276d9f81.png)

also @TheRaffy told me , that he experienced this 

some people (@pchote , @penev92 ) say that this is a old bug , but since we have our new release this appears to be more common now , though still occasionally  ... 

here is my replay where it happend on HaosRidge @ 10:34 - 10:55 ... in the "Top Left corner" ... im the light-blue player called "Wasa" : 

https://www.dropbox.com/s/h6ikq2ai21e3ztz/undeploy%20bug%20micro.orarep?dl=0

![undeployyyyyyyyy](https://cloud.githubusercontent.com/assets/9139965/7339153/848f161c-ec64-11e4-8ff5-82038b38f718.png)

**Comment by obrakmann** [comment]
I think I've found the source of the problem. The building is locked while the player gives the undeploy order, so [this check](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/Transforms.cs#L74) and the order is denied. `Building::Locked` default to `true` since 96cc1276c23f5490bd8a1a95ba2dcb11d6ce7ebd, which also made `Building` implement `InotifyBuildingComplete`. In current bleed, this implementation is gone, so `Locked` never gets set to `false`, which will be reason the bug occurs. Haven't found when and why it was removed, though.

**Comment by obrakmann** [comment]
```
(17:58:46) antares79: pchote: I assume you mean it's a race condition? sth. in RenderBuilding overwrites the DefaultAnimation before the action has a chance to run?
(18:05:20) pchote: antares79: yeah
(18:36:57) antares79: pchote: ok, thanks. I thought that having game logic be run from the rendering parts was kinda nasty
(18:38:46) pchote: antares79: for what its worth, I tried to trick penev into fixing this yesterday
(18:40:27) pchote: I've been steering his upgradeable rendering stuff to cover an issue i've been planning / working on for a couple of years
(18:40:46) pchote: which is to replace our Render* subclasses with the base RenderSprites plus With*
(18:40:56) pchote: his WithBody being the next part of that
(18:41:06) pchote: that now includes an open / start sequence
(18:41:15) pchote: for buildings we can just set that to the make animation
(18:41:23) pchote: possibly via a subclass, though
(18:41:32) pchote: and then use upgrades to cover the logic half of it
(18:41:38) pchote: which we already have an issue about
(18:42:27) pchote: we could perhaps even do this without the integrated make animation if we wanted to use two different With* sequences
(18:42:32) pchote: and have them switch based on the upgrade
```

**Comment by MicroJOo** [comment]
ive noticed another thing ... just right now ingame ... 

this particular "construction yard" ... cant undeploy nor can it "hold" finished buildings ... 

i had 2 construction conyards (and a crono to shift stuff) ... my plan was to undeploy 1 of the construction conyards , and shift the mcv to another place ... but ... i couldnt undploy one of them ... so i just undeployed the other one (that worked) ... but funny thing , the buildings i allready had finished (i did hold a AA and a barrack ... just got canceled when i undeployed the "working" construction conyard ... although i had a 2nd construction yard deployed (the bugged one)

**Comment by fgdzfggsfgsfgsf** [comment]
Yeah you can't build with the buggy Cyards either, definitely annoying.  I think you can still build within their radius though if you have another functional cyard.

**Comment by StexoO** [comment]
release 2015-04-24
You can't pack the buggy construction yards, but sometimes you can't even build buildings.
![sans titre](https://cloud.githubusercontent.com/assets/8930382/7542439/04e7e8aa-f5bd-11e4-93e4-3383fa1f1f0a.png)
Replay : http://s000.tinyupload.com/index.php?file_id=00775649103957765458
Isn't it something like "Can't build anymore if the last construction yard is destroyed" ?
Like the last MCV built is the only one who can't build buildings.

**Comment by fgdzfggsfgsfgsf** [comment]
Yeah I've never seen 2 buggy MCVs for one player, sucks for you, you could have won that game.

**Comment by obrakmann** [comment]
There's a fix now at #8402. In the mean time, don't place buildings while deploying an MCV and you'll be fine.
